Source/Particles/Resampling/VelocityCoincidenceThinning.cpp

```diff
@@ -0,0 +1,272 @@
+            amrex::ParticleReal total_weight = 0._prt, total_energy = 0._prt;
+#if !defined(WARPX_DIM_1D_Z)
+            amrex::ParticleReal cluster_x = 0._prt;
+#endif
+#if defined(WARPX_DIM_3D)
+            amrex::ParticleReal cluster_y = 0._prt;
+#endif
+            amrex::ParticleReal cluster_z = 0._prt;
+            amrex::ParticleReal cluster_ux = 0._prt, cluster_uy = 0._prt, cluster_uz = 0._prt;
+
+            // Finally, loop through the particles in the cell and merge
+            // ones in the same momentum bin
+            for (int i = cell_start; i < cell_stop; ++i)
+            {
+                particles_in_bin += 1;
+                const auto part_idx = indices[sorted_indices_data[i]];
+
+#if !defined(WARPX_DIM_1D_Z)
+                cluster_x += w[part_idx]*x[part_idx];
+#endif
+#if defined(WARPX_DIM_3D)
+                cluster_y += w[part_idx]*y[part_idx];
+#endif
+                cluster_z += w[part_idx]*z[part_idx];
+                cluster_ux += w[part_idx]*ux[part_idx];
+                cluster_uy += w[part_idx]*uy[part_idx];
+                cluster_uz += w[part_idx]*uz[part_idx];
+                total_weight += w[part_idx];
+                total_energy += w[part_idx] * Algorithms::KineticEnergy(
+                    ux[part_idx], uy[part_idx], uz[part_idx], mass
+                );
+
+                // check if this is the last particle in the current momentum bin
+                if (
+                    (i == cell_stop - 1)
+                    || (momentum_bin_number_data[sorted_indices_data[i]] != momentum_bin_number_data[sorted_indices_data[i + 1]])
+                ) {
+                    // check if the bin has more than 2 particles in it
+                    if ( particles_in_bin > 2 && total_weight > std::numeric_limits<amrex::ParticleReal>::min() ){
+                        // get average quantities for the bin
+#if !defined(WARPX_DIM_1D_Z)
+                        cluster_x /= total_weight;
+#endif
+#if defined(WARPX_DIM_3D)
+                        cluster_y /= total_weight;
+#endif
+                        cluster_z /= total_weight;
+                        cluster_ux /= total_weight;
+                        cluster_uy /= total_weight;
+                        cluster_uz /= total_weight;
+
+                        // perform merging of momentum bin particles
+                        auto u_perp2 = cluster_ux*cluster_ux + cluster_uy*cluster_uy;
+                        auto u_perp = std::sqrt(u_perp2);
+                        auto cluster_u_mag2 = u_perp2 + cluster_uz*cluster_uz;
+                        auto cluster_u_mag = std::sqrt(cluster_u_mag2);
+
+                        // calculate required velocity magnitude to achieve
+                        // energy conservation
+                        auto v_mag2 = total_energy / total_weight * (
+                            (total_energy / total_weight + 2._prt * mass * c2 )
+                            / (mass * mass * c2)
```

Comment:
What happens here with photons where `mass = 0`?

Reply:
Yeah, this is not set up to handle photons. The paper that Remi mentioned before (https://www.sciencedirect.com/science/article/abs/pii/S0010465515000405) discusses the proper handling of photons but we can leave that to a future PR.

Reply:
Ok. Until then, can an assert be added to check for that?